cmd/cdi-importer/importer.go

```diff
@@ -121,51 +106,90 @@ func main() {
 		os.Exit(1)
 	}
 	if source == controller.SourceNone {
-		if contentType == string(cdiv1.DataVolumeKubeVirt) {
-			createBlankImage(imageSize, availableDestSpace, preallocation, volumeMode, filesystemOverhead)
-			preallocationApplied = preallocation
-		} else {
-			errorEmptyDiskWithContentTypeArchive()
+		err := handleEmptyImage(contentType, imageSize, availableDestSpace, preallocation, volumeMode, filesystemOverhead)
+		if err != nil {
+			klog.Errorf("%+v", err)
+			os.Exit(1)
+		}
+
+	} else {
+		exitCode := handleImport(source, contentType, volumeMode, imageSize, filesystemOverhead, preallocation)
+		if exitCode != 0 {
+			os.Exit(exitCode)
 		}
+	}
+}
+
+func handleEmptyImage(contentType string, imageSize string, availableDestSpace int64, preallocation bool, volumeMode v1.PersistentVolumeMode, filesystemOverhead float64) error {
+	var preallocationApplied bool
+
+	if contentType == string(cdiv1.DataVolumeKubeVirt) {
+		createBlankImage(imageSize, availableDestSpace, preallocation, volumeMode, filesystemOverhead)
+		preallocationApplied = preallocation
 	} else {
-		waitForReadyFile()
-		klog.V(1).Infoln("begin import process")
+		errorEmptyDiskWithContentTypeArchive()
+	}
+
+	err := importCompleteTerminationMessage(preallocationApplied)
+	return err
+}
+
+func handleImport(
+	source string,
+	contentType string,
+	volumeMode v1.PersistentVolumeMode,
+	imageSize string,
+	filesystemOverhead float64,
+	preallocation bool) int {
+	klog.V(1).Infoln("begin import process")
 
-		ds = newDataSource(source, contentType, volumeMode)
-		defer ds.Close()
+	ds := newDataSource(source, contentType, volumeMode)
+	defer func() {
```

Comment:
Why did you put the close inside a func and not like before defer ds.Close()?

Reply:
Only this part opens dataSource, so it should also close it. By doing it here I am sure that all returns from the handleImport are covered. 

Of course while doing that, I have discovered that util.WriteTerminationMessage MUST happen before ds.close().

Reply:
Oh, now I see, you are not asking why I did the whole thing. But why there is func wrapper?
Just a leftover from tests and debugging. Will remove now.

Reply:
done